File: 1/2021.01.07/76.Minimum Window Substring/java/solution.java
class Solution {

    public String minWindow(String s, String t) {
        Map<Character, Integer> need = new HashMap<>();
        Map<Character, Integer> window = new HashMap<>();
        for (char c : t.toCharArray()) {
            need.put(c, need.getOrDefault(c, 0) + 1);
        }
        int left = 0, right = 0;
        int start = 0, valid = 0;
        int len = s.length() + 1;
        while (right < s.length()) {
            char c = s.charAt(right);
            right++;
            if (need.containsKey(c)) {
                window.put(c, window.getOrDefault(c, 0) + 1);
                // 注意：不能用 == 比较两个Integer包装类
                if (window.get(c).intValue() == need.get(c).intValue()) {
                    valid++;
                }
            }
            while (valid == need.size()) {
                if (right - left < len) {
                    start = left;
                    len = right - left;
                }
                c = s.charAt(left);
                left++;
                if (need.containsKey(c)) {
                    if (window.get(c).intValue() == need.get(c).intValue()) {
                        valid--;
                    }
                    window.put(c, window.get(c) - 1);
                }
            }
        }
        return len == s.length() + 1 ? "" : s.substring(start, start + len);

    }

}

File: 1/2021.01.07/76.Minimum Window Substring/python/solution.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        need = {}
        window = {}
        for c in t:
            need[c] = need.get(c, 0) + 1
        left = right = 0
        start = valid = 0
        LEN = len(s) + 1
        while right < len(s):
            c = s[right]
            right += 1
            if c in need:
                window[c] = window.get(c, 0) + 1
                if window[c] == need[c]:
                    valid += 1
            while valid == len(need):
                if right - left < LEN:
                    start = left
                    LEN = right - left
                c = s[left]
                left += 1
                if c in need:
                    if window[c] == need[c]:
                        valid -= 1
                    window[c] -= 1
        return '' if LEN == len(s) + 1 else s[start:start+LEN]
